### ISSUE TITLE
[EPCO-2] Compile program requirements

### ISSUE THREAD

**Issue by hatfield-c** [hatfield-c/epco#2]
This issue will be used to coordinate and compile the required functionality and requisite data structures for EPCO application.

**Comment by hatfield-c** [comment]
Calculate separation field of a neural network for a given finite domain
- A separation field D is an array of size si ~~ S = { s1 x s2 x ... x si x ... x sd }, where si is the number of uniform samples taken for dimension i.
- The entries of D correspond to points in the smooth problem space P which has dimensionality R^d. Each entry of D is a sample from P, which is fed through neural network N.
- Required data:
    - Dimensions of uniform samples to take from parameters of P.
    - Space P
    - Neural network N

**Comment by hatfield-c** [comment]
Render two axes of a separation field
- A separation field will need to be rendered as an image with dimensions that might not match the dimensions of the field.
    - Use cv2 nearest neighbor scaling to turn a numpy array of the samples into the final output image
- Required data:
    - Dimensions of separation field D
    - Dimension of desired output image
    - Separation field D

**Comment by hatfield-c** [comment]
A neural network model class
- Composed only of dense layers followed by ReLU activation layers
    - Final dense layer is not followed by ReLU layer
    - Final dense layer has one output
- There will likely be more needed methods related to search i.e. gradients

**Comment by hatfield-c** [comment]
Loss function must be specified
- Sigmoid is non-linear, and thus the usage of classical binary cross entropy would break with the "polygonal/planar/linear approximation" approach of the neural network that is meant to motivate this paper.
- A custom loss function that specifically holds this interpretation would be preferred - particularly if it minimizes false positives (false valid predictions). However this would take ample time to develop.
- A simple linear "cross entropy"-esque loss function would be:
    - y * min(1, max(0, x)) + (1 - y)(1 - min(1, max(0, x)))
    - However, this loss function suffers the "corners" problem, and is not as well understood by me
- A classical binary cross entropy loss will have to be settled upon - much of the interpretation remains intact, we have simply smoothed out the corners of the loss function. Furthermore, sigmoid is not present in N itself - it is simply used for the training process. The resulting N after training exists independently of the sigmoid activation.

**Comment by hatfield-c** [comment]
Search actions and states must be defined
- A search algorithm will need to be settled on and/or defined
    - A* is simple, but requires a lot of memory (which might be acceptable)
    - IDA* is supposed to work well with changing environments
    - RBFS (Recursive Best First Search?) can potentially be extended to include more than one alternative state?
- Special care must be paid to how states are defined during the search procedure
- Heuristic function will need to use N in some way
    - The working theory is that the gradient of N at certain positions will point towards the closest wall.
    - The previous would imply that the Hessian of N could also guide search, as it would tell us how the direction to the nearest wall is changing, from which we can make inferences about the local structure.

**Comment by hatfield-c** [comment]
An event simulator/environment must exist to work within
- The simulation can be subdivided and categorized:
    - Set E (Environmental Definition) : The simulation of the entire environment that A is contained within. Thus A entails B, and A may equal B.
    - Set A (Partitioned Variables) : The simulation values which are encoded and optimized over in N.
-  Time is discrete. Must specify how dimensional changes over time is stored and handled.

**Comment by hatfield-c** [comment]
Data loading/saving/formatting
- Training data must be loaded
- Output data will need to be saved
- Formatting rules will need to be specified for representing the data that is loaded

**Comment by hatfield-c** [comment]
This should be enough for now to get started. More issues can be created as needed.
